File: implementations/javascript/hash_table.js
const SIZES = { small: 100, medium: 10000, large: 100000 };
const TABLE_SIZE = 150001;
const n = SIZES[process.argv[2]];

const keys     = new Float64Array(TABLE_SIZE);
const vals     = new Float64Array(TABLE_SIZE);
const occupied = new Uint8Array(TABLE_SIZE);

function hashFn(key) {
    return Number(BigInt(key) * 2654435761n % BigInt(TABLE_SIZE));
}

function insert(key, val) {
    let i = hashFn(key);
    while (occupied[i]) {
        if (keys[i] === key) { vals[i] = val; return; }
        i = (i + 1) % TABLE_SIZE;
    }
    keys[i] = key;
    vals[i] = val;
    occupied[i] = 1;
}

function lookup(key) {
    let i = hashFn(key);
    while (occupied[i]) {
        if (keys[i] === key) return vals[i];
        i = (i + 1) % TABLE_SIZE;
    }
    return -1;
}

for (let i = 0; i < n; i++) insert(i * 7 + 3, i * 13 + 5);
let checksum = 0;
for (let i = 0; i < n; i++) checksum += lookup(i * 7 + 3);
console.log(checksum);

File: implementations/rust/hash_table.rs
use std::env;

const SMALL: usize      = 100;
const MEDIUM: usize     = 10000;
const LARGE: usize      = 100000;
const TABLE_SIZE: usize = 150001;

static mut KEYS: [i64; TABLE_SIZE] = [0; TABLE_SIZE];
static mut VALS: [i64; TABLE_SIZE] = [0; TABLE_SIZE];
static mut OCCUPIED: [bool; TABLE_SIZE] = [false; TABLE_SIZE];

fn hash(key: i64) -> usize {
    ((key as u64).wrapping_mul(2654435761) % TABLE_SIZE as u64) as usize
}

unsafe fn table_clear() {
    for i in 0..TABLE_SIZE { OCCUPIED[i] = false; }
}

unsafe fn insert(key: i64, val: i64) {
    let mut i = hash(key);
    while OCCUPIED[i] {
        if KEYS[i] == key { VALS[i] = val; return; }
        i = (i + 1) % TABLE_SIZE;
    }
    KEYS[i] = key;
    VALS[i] = val;
    OCCUPIED[i] = true;
}

unsafe fn lookup(key: i64) -> i64 {
    let mut i = hash(key);
    while OCCUPIED[i] {
        if KEYS[i] == key { return VALS[i]; }
        i = (i + 1) % TABLE_SIZE;
    }
    -1
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let n = match args[1].as_str() {
        "small"  => SMALL,
        "medium" => MEDIUM,
        "large"  => LARGE,
        _ => { eprintln!("Unknown size: {}", args[1]); return; }
    };
    unsafe {
        table_clear();
        for i in 0..n {
            insert(i as i64 * 7 + 3, i as i64 * 13 + 5);
        }
        let mut checksum: i64 = 0;
        for i in 0..n {
            checksum += lookup(i as i64 * 7 + 3);
        }
        println!("{}", checksum);
    }
}
